examples/chef/common/chef-rvc-mode-delegate.cpp

```diff
@@ -0,0 +1,210 @@
+    {
+        response.status = to_underlying(ModeBase::StatusCode::kGenericFailure);
+        response.statusText.SetValue(chip::CharSpan::fromCharString("Change to the mapping mode is only allowed from idle"));
+        return;
+    }
+
+    response.status = to_underlying(ModeBase::StatusCode::kSuccess);
+}
+
+CHIP_ERROR RvcRunModeDelegate::GetModeLabelByIndex(uint8_t modeIndex, chip::MutableCharSpan & label)
+{
+    if (modeIndex >= ArraySize(kModeOptions))
+    {
+        return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED;
+    }
+    return chip::CopyCharSpanToMutableCharSpan(kModeOptions[modeIndex].label, label);
+}
+
+CHIP_ERROR RvcRunModeDelegate::GetModeValueByIndex(uint8_t modeIndex, uint8_t & value)
+{
+    if (modeIndex >= ArraySize(kModeOptions))
+    {
+        return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED;
+    }
+    value = kModeOptions[modeIndex].mode;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR RvcRunModeDelegate::GetModeTagsByIndex(uint8_t modeIndex, List<ModeTagStructType> & tags)
+{
+    if (modeIndex >= ArraySize(kModeOptions))
+    {
+        return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED;
+    }
+
+    if (tags.size() < kModeOptions[modeIndex].modeTags.size())
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    std::copy(kModeOptions[modeIndex].modeTags.begin(), kModeOptions[modeIndex].modeTags.end(), tags.begin());
+    tags.reduce_size(kModeOptions[modeIndex].modeTags.size());
+
+    return CHIP_NO_ERROR;
+}
+
+ModeBase::Instance * RvcRunMode::Instance()
+{
+    return gRvcRunModeInstance;
+}
+
+void RvcRunMode::Shutdown()
+{
+    if (gRvcRunModeInstance != nullptr)
+    {
+        delete gRvcRunModeInstance;
+        gRvcRunModeInstance = nullptr;
+    }
+    if (gRvcRunModeDelegate != nullptr)
+    {
+        delete gRvcRunModeDelegate;
+        gRvcRunModeDelegate = nullptr;
+    }
+}
+
+void emberAfRvcRunModeClusterInitCallback(chip::EndpointId endpointId)
+{
+    VerifyOrDie(endpointId == 1); // this cluster is only enabled for endpoint 1.
+    VerifyOrDie(gRvcRunModeDelegate == nullptr && gRvcRunModeInstance == nullptr);
+    gRvcRunModeDelegate = new RvcRunMode::RvcRunModeDelegate;
+    gRvcRunModeInstance =
+        new ModeBase::Instance(gRvcRunModeDelegate, 0x1, RvcRunMode::Id, chip::to_underlying(RvcRunMode::Feature::kOnOff));
+    gRvcRunModeInstance->Init();
+}
+#endif // ZCL_USING_RVC_RUN_MODE_CLUSTER_SERVER
+
+#ifdef ZCL_USING_RVC_CLEAN_MODE_CLUSTER_SERVER
+#include <chef-rvc-mode-delegate.h>
+using namespace chip::app::Clusters::RvcCleanMode;
+static RvcCleanModeDelegate * gRvcCleanModeDelegate = nullptr;
+static ModeBase::Instance * gRvcCleanModeInstance   = nullptr;
+
+CHIP_ERROR RvcCleanModeDelegate::Init()
+{
+    return CHIP_NO_ERROR;
+}
+
+void RvcCleanModeDelegate::HandleChangeToMode(uint8_t NewMode, ModeBase::Commands::ChangeToModeResponse::Type & response)
+{
+    uint8_t rvcRunCurrentMode = gRvcRunModeInstance->GetCurrentMode();
+
+    if (rvcRunCurrentMode == RvcRunMode::ModeCleaning)
```

Comment:
In this example app how does the app ever get out of cleaning mode? Sure it cannot be set by a controller with the logic you have here, but I don't see where the application ever goes from cleaning back to idle after it done cleaning.

This is likely another one of those question where if there is some other buisness/app logic outside of this code that takes care of this, feel free to mark this as resolved without explaination. But if you have a few mins to help me understand how this works I would appreciate it.

Reply:
Hi @tehampson ,
   I think the calling path is from  [Instance::InvokeCommand](https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/mode-base-server/mode-base-server.cpp#L279-L290)
```
    case ModeBase::Commands::ChangeToMode::Id:
        ChipLogDetail(Zcl, "ModeBase: Entering handling ChangeToModeWithStatus");

        HandleCommand<Commands::ChangeToMode::DecodableType>(
            handlerContext, [this](HandlerContext & ctx, const auto & commandData) { HandleChangeToMode(ctx, commandData); });
```
calling to [Instance::HandleChangeToMode](https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/mode-base-server/mode-base-server.cpp#L365)
```
void Instance::HandleChangeToMode(HandlerContext & ctx, const Commands::ChangeToMode::DecodableType & commandData)
{
...
    mDelegate->HandleChangeToMode(newMode, response);
...
}
```
then calling to the real delegate's `HandleChangeMode` function.

In brief, the external controller can still call `ModeBaseCluster`'s `ChangeToMode` command to trigger the state change.

![image](https://github.com/project-chip/connectedhomeip/assets/7869096/8abf0ebe-7849-4301-ac41-cad88abfb03c)

Reply:
But isn't `mDelegate->HandleChangeToMode(newMode, response);` calling this function here which will block any change back into idle? You over overridden the base delegate

`void HandleChangeToMode(uint8_t mode, ModeBase::Commands::ChangeToModeResponse::Type & response) override;`

Reply:
@erwinpan1 ^^ (forgot to mention you in my reply)

Reply:
@tehampson  The functions above only blocked calling `RvcCleanModeDelegate::HandleChangeToMode`  when  current Running mode is doing clean (`rvcRunCurrentMode == RvcRunMode::ModeCleaning` ).  However, we should be able to call `RvcRunModeDelegate::HandleChangeToMode` to set Running mode to `RvcRunMode::ModeIdle`, right?

Reply:
In other words, you can not force the RVC to switch cleaning mode from `ModeVacuum` to `ModeWash`. But you are still able to  ask the RVC to switch running mode from `ModeCleaning` to `ModeIdle` .

Reply:
Thanks for explaining that. LGTM